Improvements to the Backend customization introduction (#1698)
strapi/documentation: docusaurus/docs/dev-docs/backend-customization.md

6. The code executed by the controllers and services interacts with the [models](/dev-docs/backend-customization/models) that are a representation of the content data structure stored in the database.<br />Interacting with the data represented by the models is handled by the [Entity Service](/dev-docs/api/entity-service) and [Query Engine](/dev-docs/api/query-engine).
7. The server returns a [response](/dev-docs/backend-customization/requests-responses#response). The response can travel back through route middlewares and global middlewares before being sent.

Both global and route middlewares include an asynchronous callback function, `await next()`. Depending on what is returned by the middleware, the request will either go through a shorter or longer path through the backend:

* If a middleware returns nothing, the request will continue travelling through the various core elements of the backend (i.e., controllers, services, and the other layers that interact with the database).
* If a middleware returns something, a response will be immediately sent, skipping the rest of the core elements.

```mermaid

> graph TB

Reviewer: This graph is incorrect.

on every layer you except for policy's you could return before calling the next faction inline.
so for example giving a policy error means that global middleware still get ran.
Or if you where to return before calling query engine in entity service it will take the same route down as it came up.
So if you return before calling next with the 4th decorator decorator 3s next would be called and then go back down the chain.
Will make a pr on top of this one with a more accurate route.

Author: cc @derrickmehaffy since the content of this graph comes from some material Derrick sent me and that was presented during a best practices showcase.

---

# Back-end customization

> :::strapi Disambiguation: Strapi back end

Author: We write "back end" when used as a noun and "back-end" as used as an adjective, as recommended [here](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/back-end).
I probably let a few slip in and will fix them. Thank you!

- The front-end part of Strapi is called the admin panel. The admin panel presents a graphical user interface to help you structure and manage the content.

Throughout this developer documentation, 'backend' refers exclusively to the back-end part of Strapi.

The [User Guide](/user-docs/intro) explains how to use the admin panel and the [admin panel customization section](/dev-docs/admin-panel-customization) details the various customization options available for the admin panel.
:::

The Strapi back end runs an HTTP server based on [Koa](https://koajs.com/), a back-end JavaScript framework.

Like any HTTP server, the Strapi back end receives requests and send responses. You can send requests to the Strapi back end to create, retrieve, update, or delete data through the [REST](/dev-docs/api/rest) or [GraphQL](/dev-docs/api/graphql) APIs.

A request can travel through the Strapi backend as follows:

1. The Strapi server receives a [request](/dev-docs/backend-customization/requests-responses#requests).
2. The request hits [global middlewares](/dev-docs/backend-customization/middlewares) that are run in a sequential order.

> 3. The request hits a [route](/dev-docs/backend-customization/routes).<br/>By default, Strapi generates routes for all the content-types that you create (see [REST API documentation](/dev-docs/api/rest)), and more routes can be added and configured.

Author: Makes sense! Thank you.

4. [Route policies](/dev-docs/backend-customization/policies) act as a read-only validation step that can block access to a route. [Route middlewares](/dev-docs/backend-customization/routes#middlewares) can control the request flow and mutate the request itself before moving forward.
5. [Controllers](/dev-docs/backend-customization/controllers) execute code once a route has been reached. [Services](/dev-docs/backend-customization/services) are optional, additional code that can be used to build custom logic reusable by controllers.
6. The code executed by the controllers and services interacts with the [models](/dev-docs/backend-customization/models) that are a representation of the content data structure stored in the database.<br />Interacting with the data represented by the models is handled by the [Entity Service](/dev-docs/api/entity-service) and [Query Engine](/dev-docs/api/query-engine).
7. The server returns a [response](/dev-docs/backend-customization/requests-responses#response). The response can travel back through route middlewares and global middlewares before being sent.

> Both global and route middlewares include an asynchronous callback function, `await next()`. Depending on what is returned by the middleware, the request will either go through a shorter or longer path through the backend:

Reviewer: ```suggestion
Both global and route middlewares include an asynchronous callback function, `await next()`. Depending on when you return the middleware, the request will either go through a shorter or longer path through the backend:
```

Reviewer: what you return is for policies not middleware aka
true to continue
false to give a policy error
or policy error for a custom policy error

> * If a middleware returns nothing, the request will continue travelling through the various core elements of the backend (i.e., controllers, services, and the other layers that interact with the database).

Reviewer: ```suggestion
* caling `await next()` means it continues travelling through the various core elements of the backend (i.e., controllers, services, and the other layers that interact with the database). what this middleware returns is what the middleware above it will get as its `await next()` output.
```

> * If a middleware returns something, a response will be immediately sent, skipping the rest of the core elements.

Reviewer: ```suggestion
* If a middleware returns before calling `await next()`, a response will be immediately sent, skipping the rest of the core elements. Then it will go back down the same chain it came up.
```
